contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,483 @@
+        ) -> Result<Result<(), PSP22Error>, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_SELECTOR))
+                        .push_arg(to)
+                        .push_arg(amount)
+                        .push_arg(vec![0x0]),
+                )
+                .returns::<Result<(), PSP22Error>>()
+                .fire()
+        }
+
+        /// Transfers a given amount of a PSP22 token on behalf of a specified account to another account
+        ///
+        /// Will revert if not enough allowance was given to the caller prior to executing this tx
+        fn transfer_from_tx(
+            &self,
+            token: AccountId,
+            from: AccountId,
+            to: AccountId,
+            amount: Balance,
+        ) -> Result<Result<(), PSP22Error>, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_FROM_SELECTOR))
+                        .push_arg(from)
+                        .push_arg(to)
+                        .push_arg(amount)
+                        .push_arg(vec![0x0]),
+                )
+                .call_flags(CallFlags::default().set_allow_reentry(true)) // needed for checking allowance before the actual tx
+                .returns::<Result<(), PSP22Error>>()
+                .fire()
+        }
+
+        /// Returns the amount of unused allowance that the token owner has given to the spender
+        fn allowance(
+            &self,
+            token: AccountId,
+            owner: AccountId,
+            spender: AccountId,
+        ) -> Result<Balance, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(ALLOWANCE_SELECTOR))
+                        .push_arg(owner)
+                        .push_arg(spender),
+                )
+                .returns::<Balance>()
+                .fire()
+        }
+
+        /// Returns DEX balance of a PSP22 token for an account
+        fn balance_of(&self, token: AccountId, account: AccountId) -> Result<Balance, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(BALANCE_OF_SELECTOR)).push_arg(account),
+                )
+                .returns::<Balance>()
+                .fire()
+        }
+
+        /// Swap trade output given a curve with equal token weights
+        ///
+        /// swap_fee_percentage (integer) is a percentage of the trade that goes towards the pool
+        /// B_0 - (100 * B_0 * B_i) / (100 * (B_i + A_i) -A_i * fee)
```

Comment:
Just FYI, apparently Balancer deals with 0.0001 fee precision: https://docs.balancer.fi/getting-started/faqs/fees#what-are-the-fees-for-a-trade